atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueServiceTest.java

```diff
@@ -89,6 +96,144 @@ public void canConcurrentlyReadFromStableTablesWhileCreatingAndDroppingOtherTabl
+        // Simulate a pair of KVS clients hitting the same raw KVS
+        KvTableMappingService tableMapper1 = KvTableMappingService.createWithCustomNamespaceValidation(
+                rawKvs,
+                timestamp::incrementAndGet,
+                Namespace.STRICTLY_CHECKED_NAME,
+                new TableMappingCacheConfiguration() {
+                    @Override
+                    public Predicate<TableReference> cacheableTablePredicate() {
+                        return tableReference -> cacheTableLookups;
+                    }
+                });
+        KvTableMappingService tableMapper2 = KvTableMappingService.createWithCustomNamespaceValidation(
+                rawKvs,
+                timestamp::incrementAndGet,
+                Namespace.STRICTLY_CHECKED_NAME,
+                new TableMappingCacheConfiguration() {
+                    @Override
+                    public Predicate<TableReference> cacheableTablePredicate() {
+                        return tableReference -> cacheTableLookups;
+                    }
+                });
+        TableRemappingKeyValueService kvs1 = TableRemappingKeyValueService.create(rawKvs, tableMapper1);
+        TableRemappingKeyValueService kvs2 = TableRemappingKeyValueService.create(rawKvs, tableMapper2);
+        // Create the table in the first node
+        kvs1.createTable(DATA_TABLE_REF, AtlasDbConstants.GENERIC_TABLE_METADATA);
+        Cell cell = Cell.create(PtBytes.toBytes("row"), PtBytes.toBytes("col"));
+        // Try to get a value, creating the mapping in memory
+        kvs1.get(DATA_TABLE_REF, ImmutableMap.of(cell, 1L));
+        // Drop the table in node 2. This removes the mapping from the db, but the in memory mapping in node 1 isn't
+        // removed
+        kvs2.dropTable(DATA_TABLE_REF);
+        // Create the table again in node 2. This will write a new mapping to the db, which doesn't match the mapping
+        // that is still in memory on node 1.
+        kvs2.createTable(DATA_TABLE_REF, AtlasDbConstants.GENERIC_TABLE_METADATA);
+        // Write some data to kvs2 and try to read it back from kvs1. This verifies that the table caches across 1 & 2
+        // are consistent.
+        kvs2.put(DATA_TABLE_REF, ImmutableMap.of(cell, new byte[] {0x42}), 1L);
+        // Try and get a value on both nodes and make sure they are the same.
+        Map<Cell, Value> result = kvs1.get(DATA_TABLE_REF, ImmutableMap.of(cell, 2L));
+        Map<Cell, Value> result2 = kvs2.get(DATA_TABLE_REF, ImmutableMap.of(cell, 2L));
+        assertThat(result).isNotEmpty();
+        assertThat(result2).isNotEmpty();
+        assertThat(result).isEqualTo(result2);
+    }
+
+    @Test
+    @Disabled("Issues with both table creation and metadata storage after table creation exist to cause this test to"
+            + " fail.")
+    public void fuzzTestConcurrentlyCreateAndDeleteTablesAcrossManyWriters() {
```

Comment:
This is a bit of a torture test but reliably reproduces the problems even with caching disabled, surfacing other issues with the table creation/deletion workflow and with KvTableMappingService.